consensus/event_loop.go

```diff
@@ -544,6 +574,121 @@ func (fc *FastHotStuffEventLoop) tryConstructVoteQCInCurrentView() (
+		if _, ok := validatorSetAtBlock[timeout.GetPublicKey().ToString()]; !ok {
+			continue
+		}
+
+		// Update the highQC if the timeout message has a higher QC view than the current highQC's view
+		if isInterfaceNil(validatorsHighQC) || timeout.GetHighQC().GetView() > validatorsHighQC.GetView() {
+			validatorsHighQC = timeout.GetHighQC()
+		}
+	}
+
+	// If we didn't find a high QC or didn't find any valid timeout messages, then we can't build a timeout QC.
+	if isInterfaceNil(validatorsHighQC) {
+		return false, nil, nil, nil, nil, nil
+	}
+
+	// Fetch the validator set for the block height of the high QC. This lookup is guaranteed to succeed
+	// because it succeeded above.
+	ok, previousBlock, validatorSet, _ := fc.getBlockAndValidatorSetByHash(validatorsHighQC.GetBlockHash())
+	if !ok {
+		return false, nil, nil, nil, nil, nil
+	}
+
+	// Compute the total stake and total stake with timeouts
+	totalStake := uint256.NewInt()
+	totalTimedOutStake := uint256.NewInt()
+
+	// Track the high QC view for each validator
+	highQCViews := make([]uint64, len(validatorSet))
+
+	// Track the signatures and signers list for validators who timed out
+	signersList := bitset.NewBitset()
+	signatures := []*bls.Signature{}
+
+	// Iterate through the entire validator set and check if each one has timed out for the previous
+	// view. Track all validators who timed out and their stakes. We iterate through the validator set
+	// here rather than the timeoutsByValidator map because we want to preserve the order of the validator
+	// for the signersList bitset. In practice, the validator set is expected to be <= 1000 in size, so
+	// this loop will be fast.
+	for ii, validator := range validatorSet {
+		totalStake = uint256.NewInt().Add(totalStake, validator.GetStakeAmount())
+
+		// Skip the validator if it hasn't timed out for the previous view
+		timeout, hasTimedOut := timeoutsByValidator[validator.GetPublicKey().ToString()]
+		if !hasTimedOut {
+			continue
+		}
+
+		// Compute the signature payload that the validator should have signed
+		signaturePayload := GetTimeoutSignaturePayload(timeout.GetView(), timeout.GetHighQC().GetView())
+
+		// Verify the timeout signature
+		if !isValidSignatureSinglePublicKey(timeout.GetPublicKey(), timeout.GetSignature(), signaturePayload[:]) {
+			continue
+		}
+
+		// Track the signatures, timed out stake, and high QC views for the validator
+		totalTimedOutStake = uint256.NewInt().Add(totalTimedOutStake, validator.GetStakeAmount())
+		signersList.Set(ii, true)
+		signatures = append(signatures, timeout.GetSignature())
+		highQCViews[ii] = timeout.GetHighQC().GetView()
+	}
+
+	// Check if we have a super majority of stake that has timed out
+	if !isSuperMajorityStake(totalTimedOutStake, totalStake) {
+		return false, nil, nil, nil, nil, nil
+	}
+
+	// Finally aggregate the signatures from the timeouts
+	aggregateSignature, err := bls.AggregateSignatures(signatures)
```

Comment:
The documentation for AggregateSignatures needs a little update. At the end, it should say that signatures could be coming from different payloads.

Reply:
Good catch. The comment is updated now